src/OpenSage.Game/Logic/Object/Behaviors/PhysicsBehavior.cs

```diff
@@ -7,18 +9,62 @@ public sealed class PhysicsBehavior : BehaviorModule
         private readonly GameObject _gameObject;
         private readonly PhysicsBehaviorModuleData _moduleData;
 
+        // TODO: Don't know if this belongs here.
+        private Vector3 _velocity;
+
+        private Vector3 _cumulativeForces;
+
+        public float Mass { get; set; }
+
         internal PhysicsBehavior(GameObject gameObject, PhysicsBehaviorModuleData moduleData)
         {
             _gameObject = gameObject;
             _moduleData = moduleData;
+
+            Mass = moduleData.Mass;
         }
 
         internal override void Update(BehaviorUpdateContext context)
         {
+            // Calculate force due to gravity.
             var gravity = context.GameContext.AssetLoadContext.AssetStore.GameData.Current.Gravity;
-            var mass = _moduleData.Mass;
+            var gravityAcceleration = Mass * new Vector3(0, 0, gravity);
+
+            var cumulativeAcceleration = _cumulativeForces / Mass;
+            _cumulativeForces = Vector3.Zero;
+
+            var acceleration = gravityAcceleration + cumulativeAcceleration;
+
+            // Integrate velocity.
+            var deltaTime = (float) context.Time.DeltaTime.TotalSeconds;
+            _velocity += acceleration * deltaTime;
+
+            // Integrate position.
+            var newTranslation = context.GameObject.Transform.Translation + (_velocity * deltaTime);
 
-            // TODO
+            var terrainHeight = context.GameContext.Terrain.HeightMap.GetHeight(
+                newTranslation.X,
+                newTranslation.Y);
+
+            if (newTranslation.Z < terrainHeight)
+            {
+                newTranslation.Z = terrainHeight;
+
+                // TODO: Improve bouncing.
+                _velocity.Z = Math.Abs(_velocity.Z) * 0.9f;
```

Comment:
Why you multiply this with 0.9 here? I guess this is just a placeholder?

Reply:
Yes, it’s a placeholder. I didn’t want it to bounce forever. I haven’t yet looked into how it’s supposed to work.